src/nupic/algorithms/SDRClassifier.cpp

```diff
@@ -0,0 +1,745 @@
+/* ---------------------------------------------------------------------
+ * Numenta Platform for Intelligent Computing (NuPIC)
+ * Copyright (C) 2016, Numenta, Inc.  Unless you have an agreement
+ * with Numenta, Inc., for a separate license for this software code, the
+ * following terms and conditions apply:
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU Affero Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero Public License
+ * along with this program.  If not, see http://www.gnu.org/licenses.
+ *
+ * http://numenta.org/licenses/
+ * ---------------------------------------------------------------------
+ */
+
+#include <cmath>
+#include <deque>
+#include <iostream>
+#include <limits>
+#include <map>
+#include <string>
+#include <sstream>
+#include <vector>
+#include <stdio.h>
+
+#include <capnp/message.h>
+#include <capnp/serialize.h>
+#include <kj/std/iostream.h>
+
+#include <nupic/algorithms/ClassifierResult.hpp>
+#include <nupic/algorithms/SDRClassifier.hpp>
+#include <nupic/proto/SdrClassifier.capnp.h>
+#include <nupic/math/ArrayAlgo.hpp>
+#include <nupic/utils/Log.hpp>
+
+using namespace std;
+
+namespace nupic
+{
+  namespace algorithms 
+  {
+    namespace sdr_classifier
+    {
+
+      SDRClassifier::SDRClassifier(
+          const vector<UInt>& steps, Real64 alpha, Real64 actValueAlpha,
+          UInt verbosity) : alpha_(alpha), actValueAlpha_(actValueAlpha),
+          learnIteration_(0), recordNumMinusLearnIteration_(0), 
+          maxInputIdx_(0), maxBucketIdx_(0), version_(Version), 
+          verbosity_(verbosity)
+      {
+        for (const auto& step : steps)
+        {
+          steps_.push_back(step);
+        }
+        recordNumMinusLearnIterationSet_ = false;
+        maxSteps_ = 0;
+        for (auto& elem : steps_)
+        {
+          UInt current = elem + 1;
+          if (current > maxSteps_)
+          {
+            maxSteps_ = current;
+          }
+        }
+        actualValues_.push_back(0.0);
+        actualValuesSet_.push_back(false);
```

Comment:
These would be better in the initialization list so that the vector is populated in its constructor.

`: actualValues_({0.0})`, etc.